### ISSUE TITLE
Define/Use a code convention for new/updated code

### ISSUE THREAD

**Issue by juherr** [testng-team/testng#1375]
I think we should define a code convention for new & updated code in order to have the same IDE configuration everywhere and avoid syntax typo in pull requests.

Usually, I use [the google convention](https://google.github.io/styleguide/javaguide.html) because they provide [the idea and eclipse configuration file](https://github.com/google/styleguide).

@testng-team WDYT?

**Comment by krmahadevan** [comment]
I agree. We could also perhaps include a `editorConfig` file. In case of the formatter and user would still need to import it into their IDE for it to take effect. But in the case of editorConfig, it comes bundled with IntelliJ. Please see complete list [here](http://editorconfig.org/#download)  
They also have a plugin for eclipse that can be downloaded. Am not trying to compare this against that, but just throwing out options.

**Comment by juherr** [comment]
Yes, EditorConfig is another option. But we need to determine the rules before. And I think we should not apply it to the current code quickly because the diff will be too huge.

We can discuss on rules but I don't want to spend to much time because I don't care to follow rules I don't like (once they are defined).
The google convention could be a good start and we can customize it according to opinion of @testng-team

**Comment by cbeust** [comment]
Agreed, it would be nice to have formatting rules for IDEA and Eclipse included in the project, but in my experience, it's not always easy without specifying preferences for a lot of other things in the IDE (not sure if it's still true). If it's possible to commit a file that only specifies code formatting and nothing else, great.

Also, that file should ideally define import orders so we never see imports shuffling in the PR's.

**Comment by juherr** [comment]
@cbeust And WDYT about the google style?

**Comment by cbeust** [comment]
I'm fine with the Google style.

**Comment by missedone** [comment]
> Agreed, it would be nice to have formatting rules for IDEA and Eclipse included in the project, but in my experience, it's not always easy without specifying preferences for a lot of other things in the IDE (not sure if it's still true). 

This still sucks today, for import order, eclipse has a separate preference and can be exported.
For now, our team uses both Eclipse and Intellij, we use eclipse preference as the source of truth. For Intellij, there is a Eclipse Code Formatter plugin can import the eclipse preference files (including the import order one)

**Comment by juherr** [comment]
@missedone As proposed by @krmahadevan , editorConfig is supposed to fix configuration issues. Do you have any bad experience with it?

BTW, if we follow the google style, they already provide eclipse and intellij configuration files.

**Comment by missedone** [comment]
I didn't use editorConfig before, it could be a fun with trying something new, but looks like need to define the editorConfig profile from scratch to fit TestNG...

I would say let's start with google style first.
for the import order, there need to be a separate profile file.

**Comment by juherr** [comment]
Closing this as completed.

The original goal is now covered by the current contributor tooling:

- #2590 introduced enforced Java formatting with Autostyle and Google Java Format, including import ordering.
- #2596 aligned `.editorconfig` with the project code style.
- #2629 documented the `autostyleCheck` / `autostyleApply` workflow for contributors as part of #2597.

The formatting setup is still actively used by the build; possible future changes to the formatter itself are tracked separately in #3292.
